### PR TITLE
Extend metrics middleware

### PR DESCRIPTION
**Pull request by hodgestar**
This should allow us to track via Graphite, for each transport, network operator and tag or tagpool:
- number of sessions
- number of messages
- time taken for each session
- time taken for each session, rounded to the next `N` seconds (this is to allow very rough session costs calculations since some networks bill per `N` seconds or part there-of).

See also praekelt/vumi#856.

### CODE REVIEW

File: go/vumitools/middleware.py [hunk 1]
```diff
@@ -78,28 +78,81 @@ def is_optout_message(message):
...
+    :param dict tagpools:
+        A dictionary defining which tag pools and tags should be tracked.
+        E.g.::
+
+            tagpools:
+                pool1:
+                    __track_pool__: true
+                    tagA: true
+                pool2:
+                    tagB: true
```
Comment (jerith) [inline go/vumitools/middleware.py]: I'd rather structure this differently, perhaps something like:

```
tagpools:
    pool1:
        track_pool: true
        track_tags: ["tagA"]
    pool2:
        track_pool: false
        track_tags: ["tagB"]
    pool3:
        track_pool: true
        track_all_tags: true
```

**Comment by jerith** [comment]
Despite my comment about the config format, I don't really think we should be trying to track tag-level things here. It adds additional ordering-dependence to our already-complicated middleware stack and potentially introduces exciting new failure modes.

Maybe we need something on the routing table dispatcher?

**Comment by hodgestar** [comment]
I like the suggestion for the tagpools config format. Changing to that.

**Comment by hodgestar** [comment]
I'd prefer all of this to live on the transports because:
- messages bounce through the routing table dispatcher multiple times
- we're trying to measure the length of sessions, which might be different at the routing table dispatcher and the transport (and the transport value is more correct).

File: go/vumitools/middleware.py [hunk 2]
```diff
@@ -78,28 +78,82 @@ def is_optout_message(message):
...
+    * The number of messages sent and received.
+    * The number of sessions started.
+    * The length of each session.
+
+    The network operator is determined by examining each message. If the
+    network operator is not detected by the transport, consider using network
+    operator detecting middleware to provide it.
+
+    Network operator metrics must be enabled by setting ``operator_metrics`` to
+    ``true``.
```
Comment (jerith) [inline go/vumitools/middleware.py:103]: Should we maybe rename this to `provider_metrics`? That's the field we're using in the message object and it's more inclusive of things like twitter and mxit than "network operator".

Comment (hodgestar) [inline go/vumitools/middleware.py:103]: Happy to rename to `provider_metrics`.

**Comment by hodgestar** [comment]
Ready for review.

**Comment by jerith** [comment]
> I'd prefer all of this to live on the transports because:
> - messages bounce through the routing table dispatcher multiple times

Agreed. We can use some tricks to limit this to transport connectors, but that gets ugly.

> - we're trying to measure the length of sessions, which might be different at the routing table dispatcher and the transport (and the transport value is more correct).

I was thinking that we could use the timestamps in the message objects, but that doesn't work so well for outbound messages.

I withdraw my objection, but I still wish there were a better way.

**Comment by jerith** [comment]
This middleware could really benefit from using confmodel instead of raw dicts, but we should probably make that change across all middlewares in a separate PR.

File: go/vumitools/middleware.py [hunk 3]
```diff
@@ -178,35 +258,160 @@ def set_inbound_timestamp(self, transport_name, message):
...
+                '%s.tagpool.%s' % (name, pool), msg, session_dt)
+        if config.get('track_all_tags') or tagname in config['tags']:
+            slugname = self.slugify_tagname(tagname)
+            self.fire_inbound_metrics(
+                '%s.tag.%s.%s' % (name, pool, slugname), msg, session_dt)
+
+    def fire_outbound_metrics(self, prefix, msg, session_dt):
+        self.increment_counter(prefix, 'outbound')
+        if session_dt is not None:
+            self.fire_session_dt(prefix, session_dt)
```
Comment (jerith) [inline go/vumitools/middleware.py:356]: Do we need to handle outbound messages that start sessions? I don't think we have any of those at present, but it's something that will probably happen eventually even if we don't.

**Comment by jerith** [comment]
After reviewing the changes (for which the diff view is useful), can we perhaps make this a new middleware (temporarily, at least) and revert the changes to the existing middleware? That will let us apply it to the small number of things that urgently need it with an easier rollback plan and without the risk of breaking other transports.

**Comment by hodgestar** [comment]
Without configuration changes the new middleware only publishes a new
session metric and it has tests that cover that. Not sure what the
advantage of making a new temporary middleware is.

File: go/vumitools/middleware.py (hunk 3 above)
Comment (hodgestar) [inline go/vumitools/middleware.py:356]: Thought about this and it seemed corner case, but happy to add support for
it.

**Comment by jerith** [comment]
> Without configuration changes the new middleware only publishes a new session metric and it has tests that cover that. Not sure what the advantage of making a new temporary middleware is.

The advantage is that unexpected interactions with the new code are limited to places where we explicitly use the new version rather than potentially cropping up in exotic or low-volume transports.

I'd agree with you if the changes were purely additive, but this branch also touches a bunch of existing code and therefore has the potential to break existing functionality.

**Comment by jerith** [comment]
+1 to the most recent commit.
